### PR TITLE
Address translation

### PR DESCRIPTION
**Pull request by dmitrii-ubskii**
## Usage and product changes

We allow the user to provide a translation map from the advertised server addresses (now treated as generic server names) to the actual addresses the driver shall use to connect to the cloud instances. We require the user to provide the full mapping.

Example usage:

Python:
```python
with TypeDB.cloud_driver({
            "0.deployment-UUID.cloud.typedb.com:1729": "localhost:11729",
            "1.deployment-UUID.cloud.typedb.com:1729": "localhost:21729",
            "2.deployment-UUID.cloud.typedb.com:1729": "localhost:31729"
        }, credential) as driver:
    pass
```

Rust:
```rust
Connection::new_cloud_with_translation(                                                        
    [                                                                                          
        ("0.deployment-UUID.cloud.typedb.com:1729", "localhost:11729"),
        ("1.deployment-UUID.cloud.typedb.com:1729", "localhost:21729"),
        ("2.deployment-UUID.cloud.typedb.com:1729", "localhost:31729"),
    ].into(),                                                                                  
    credential                                                                                   
)                                                                                              
```

### CODE REVIEW

File: c/swig/typedb_driver_csharp.swg [hunk 1]
```diff
@@ -620,27 +620,27 @@
...
-            arrayPtr[arraySize] = global::System.IntPtr.Zero;
+            arrayPtr[arraySize$csinput] = global::System.IntPtr.Zero;
         }",
     post="
         unsafe
         {
             global::System.IntPtr* arrayPtr = (global::System.IntPtr*)unmanaged$csinput.ToPointer();
 
-            for (int i = 0; i < arraySize; i++)
+            for (int i = 0; i < arraySize$csinput; i++)
```
Comment (dmitrii-ubskii) [inline c/swig/typedb_driver_csharp.swg:643]: This broke once we had two input arrays in a function (`connection_open_cloud_translated`). `$csinput` appends the original variable name to `arraySize`, disambiguating the two variables.

File: c/src/database.rs [hunk 2]
```diff
@@ -110,7 +110,7 @@ pub extern "C" fn replica_info_drop(replica_info: *mut ReplicaInfo) {
 /// Retrieves the address of the server hosting this replica
 #[no_mangle]
 pub extern "C" fn replica_info_get_address(replica_info: *const ReplicaInfo) -> *mut c_char {
-    release_string(borrow(replica_info).address.to_string())
+    release_string(borrow(replica_info).server_id.to_string())
 }
```
Comment (dmitrii-ubskii) [inline c/src/database.rs:114]: Should be renamed and rephrased if the Rust changes stick.

File: rust/src/common/error.rs [hunk 3]
```diff
@@ -84,7 +86,7 @@ error_messages! { InternalError
         3: "Unexpected request type for remote procedure call: {request_type}.",
     UnexpectedResponseType { response_type: String } =
         4: "Unexpected response type for remote procedure call: {response_type}.",
-    UnknownConnectionAddress { address: Address } =
+    UnknownConnectionAddress { address: String } =
         5: "Received unrecognized address from the server: {address}.",
```
Comment (dmitrii-ubskii) [inline rust/src/common/error.rs:90]: TODO

File: rust/src/common/info.rs [hunk 4]
```diff
@@ -21,12 +21,10 @@ use std::time::Duration;
 
 use tokio::sync::mpsc::UnboundedSender;
 
-use super::{address::Address, SessionID};
-use crate::common::Callback;
+use super::{Callback, SessionID};
 
 #[derive(Clone, Debug)]
 pub(crate) struct SessionInfo {
-    pub(crate) address: Address,
```
Comment (dmitrii-ubskii) [inline rust/src/common/info.rs:29]: Not necessary

File: rust/src/common/info.rs [hunk 5]
```diff
@@ -42,7 +40,7 @@ pub(crate) struct DatabaseInfo {
 #[derive(Debug)]
 pub struct ReplicaInfo {
     /// The address of the server hosting this replica
-    pub address: Address,
+    pub server_id: String,
```
Comment (dmitrii-ubskii) [inline rust/src/common/info.rs:43]: TODO docs (if this change is approved)

File: rust/src/connection/connection.rs [hunk 6]
```diff
@@ -248,22 +323,20 @@ impl fmt::Debug for Connection {
 
 #[derive(Clone)]
 pub(crate) struct ServerConnection {
-    address: Address,
```
Comment (dmitrii-ubskii) [inline rust/src/connection/connection.rs:251]: The only reason for using this was to allow re-fetching the same `ServerConnection` from the overarching `Connection` later. In most cases, the address / server_id is already available.

File: rust/src/database/database.rs [hunk 7]
```diff
@@ -178,16 +172,16 @@ impl Database {
     #[cfg_attr(feature = "sync", maybe_async::must_be_sync)]
     pub(super) async fn run_on_any_replica<F, P, R>(&self, task: F) -> Result<R>
     where
-        F: Fn(ServerDatabase, ServerConnection) -> P,
+        F: Fn(ServerDatabase) -> P,
```
Comment (dmitrii-ubskii) [inline rust/src/database/database.rs:175]: `ServerConnection` is already available through the `ServerDatabase`

File: rust/tests/integration/network.rs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+use futures::StreamExt;
+use serial_test::serial;
+use typedb_driver::{Connection, Credential, DatabaseManager, Session, SessionType::Data, TransactionType::Write};
+
+use super::common;
+
+#[tokio::test]
+#[serial]
+async fn address_translation() {
+    let connection = Connection::new_cloud_address_map(
```
Comment (flyingsilverfin) [inline rust/tests/integration/network.rs:31]: Synchronise name

File: rust/src/common/error.rs [hunk 9]
```diff
@@ -72,6 +72,8 @@ error_messages! { ConnectionError
         22: "Connection failed. Please check the server is running and the address is accessible. Encrypted Cloud endpoints may also have misconfigured SSL certificates.",
     MissingPort { address: String } =
         23: "Invalid URL '{address}': missing port.",
+    AddressTranslationMismatch { unknown: HashSet<String>, unmapped: HashSet<String> } =
+        24: "Address translation map does not match the advertised server list. Mapped servers not known to the remote server: {unknown:?}. Servers not mapped to addresses: {unmapped:?}.",
```
Comment (flyingsilverfin) [inline rust/src/common/error.rs:76]:
```
Address translation map does not match the server's advertised address list. User-provided servers not in the advertised list: {...}; Advertised servers not mapped by user: {...}
```

File: rust/src/connection/connection.rs [hunk 10]
```diff
@@ -121,14 +124,86 @@ impl Connection {
...
+    ///         "admin",
+    ///         "password",
+    ///         Some(&PathBuf::from(
+    ///             std::env::var("ROOT_CA")
+    ///                 .expect("ROOT_CA environment variable needs to be set for cloud tests to run"),
+    ///         )),
+    ///     )?,
+    /// )
+    /// ```
+    pub fn new_cloud_address_map<T, U>(address_translation: HashMap<T, U>, credential: Credential) -> Result<Self>
```
Comment (flyingsilverfin) [inline rust/src/connection/connection.rs:167]: `new_cloud_with_translation`?

**Review by flyingsilverfin — CHANGES_REQUESTED** [review]

File: .factory/automation.yml [hunk 11]
```diff
@@ -272,7 +273,7 @@ build:
         tool/test/stop-cloud-servers.sh
         exit $TEST_SUCCESS
 
-    test-python-integration-core:
+    test-python-integration:
```
Comment (dmitrii-ubskii) [inline .factory/automation.yml:276]: Unified integration tests across languages.

File: cpp/include/typedb/connection/driver.hpp [hunk 12]
```diff
@@ -83,10 +85,12 @@ class Driver {
      * Driver::cloudDriver(addresses, credential);
      * </pre>
      *
-     * @param addresses The address(es) of the TypeDB server(s)
+     * @param addresses The address(es) of the TypeDB server(s) or translation map from addresses
+     * received from the TypeDB server(s) to addresses to be used by the driver for connection
```
Comment (dmitrii-ubskii) [inline cpp/include/typedb/connection/driver.hpp:89]: Constructor overloads don't seem to get separate docs generated.

File: cpp/lib/connection/driver.cpp [hunk 13]
```diff
@@ -30,22 +30,40 @@ void Driver::initLogging() {
     _native::init_logging();
 }
 
-Driver Driver::coreDriver(const std::string& coreAddress) {
-    auto p = _native::connection_open_core(coreAddress.c_str());
+Driver Driver::coreDriver(const std::string& address) {
+    auto p = _native::connection_open_core(address.c_str());
```
Comment (dmitrii-ubskii) [inline cpp/lib/connection/driver.cpp:34]: Bring in line with declaration.

File: csharp/Drivers.cs [hunk 14]
```diff
@@ -70,12 +70,18 @@ public static ITypeDBDriver CloudDriver(string address, TypeDBCredential credent
          * Drivers.CloudDriver(addresses, credential);
          * </pre>
          *
-         * @param addresses The address(es) of the TypeDB server(s)
+         * @param addresses The address(es) of the TypeDB server(s) or translation map from addresses
+         * received from the TypeDB server(s) to addresses to be used by the driver for connection
```
Comment (dmitrii-ubskii) [inline csharp/Drivers.cs:74]: Same as C++, separate docs for overloaded ctors aren't generated.

File: python/typedb/driver.py [hunk 15]
```diff
@@ -77,5 +78,7 @@ def cloud_driver(addresses: Union[Iterable[str], str], credential: TypeDBCredent
         """
         if isinstance(addresses, str):
             return _Driver([addresses], credential)
+        elif isinstance(addresses, Mapping):
+            return _Driver(dict(addresses), credential)
```
Comment (dmitrii-ubskii) [inline python/typedb/driver.py:82]: A `Mapping` is an `Iterable`, but not vice versa, so this check must come first.

**Review by flyingsilverfin — APPROVED** [review]

**Pull request merged**